File: hackerrank/problem solving/algorithms/2 - implementation/27 - sequence equation/sequence_equation.rb
# https://www.hackerrank.com/challenges/permutation-equation/problem?isFullScreen=true

def main
  gets
  arr = read_numbers
  sequence_equation(arr) { puts _1 }
end

def read_numbers
  gets.split.map!(&:to_i)
end

# n: length of arr
# T: O(n)
# S: O(n) extra space
def sequence_equation(arr)
  values_to_index = Array.new(arr.size)
  arr.each_with_index { values_to_index[_1 - 1] = _2 }
  values_to_index.each { yield values_to_index[_1] + 1 }
end

main if __FILE__ == $0

File: hackerrank/problem solving/algorithms/2 - implementation/27 - sequence equation/sequence_equation.js
// https://www.hackerrank.com/challenges/permutation-equation/problem?isFullScreen=true

function main() {
  readLine();
  const arr = readNumbers();
  for (const x of sequenceEquation(arr)) console.log(x);
}

function readNumbers() {
  return readLine()
    .split(' ')
    .map((x) => +x);
}

// n: length of arr
// T: O(n)
// S: O(n) extra space
function* sequenceEquation(arr) {
  const valuesToIndex = Array(arr.length).fill(0);
  let i = 0;
  for (const val of arr) valuesToIndex[val - 1] = i++;
  for (const val of valuesToIndex) yield valuesToIndex[val] + 1;
}

////////////////////////////////////////////////

const readline = require('readline');

const rl = readline.createInterface({ input: process.stdin });
let inputLines = [];
rl.on('line', (line) => inputLines.push(line));
rl.on('close', main);

function readLine() {
  return inputLines.shift();
}
